### PR TITLE
Make CLUSTER SETSLOT with TIMEOUT 0 block indefinitely

### LINKED ISSUE #422: Make CLUSTER SETSLOT ... TIMEOUT 0 bypass the wait
Follow-up of #245.

TIMEOUT 0 is interpreted as the default value (2 seconds?) which is counter-intuitive given that TIMEOUT 1 is 1ms.

It's better that TIMEOUT 0 bypasses replication-before-execution.

@madolson wrote
> If a node is disconnected might be a valid use case? If a replica is unavailable for some reason, you might still want this to through. Maybe you're trying to move the data from an unhealthy to a healthy shard by migrating the data.

Original discussion: https://github.com/valkey-io/valkey/pull/245#discussion_r1587252134_

### PR DESCRIPTION
**Pull request by PingXie**
This aligns the behaviour with established Valkey commands with a TIMEOUT argument, such as BLPOP.

Fix #422

### CODE REVIEW

File: src/cluster_legacy.c [hunk 1]
```diff
@@ -5961,11 +5956,23 @@ int clusterParseSetSlotCommand(client *c, int *slot_out, clusterNode **node_out,
...
     } else {
         addReplyError(c, "Invalid CLUSTER SETSLOT action or number of arguments. Try CLUSTER HELP");
         return 0;
     }
 
+    /* Process optional arguments */
+    for (int i = optarg_pos; i < c->argc; i++) {
+        if (!strcasecmp(c->argv[i]->ptr, "timeout")) {
+            if (i + 1 >= c->argc) {
+                addReplyError(c, "Missing timeout value");
```
Comment (zuiderkwast) [inline src/cluster_legacy.c:5969]: Most checks like this simply use this shared error message:

```suggestion
                addReplyErrorObject(c, shared.syntaxerr);
```

File: tests/unit/cluster/slot-migration.tcl [hunk 2]
```diff
@@ -335,6 +335,27 @@ start_cluster 3 3 {tags {external:skip cluster} overrides {cluster-allow-replica
...
+        catch {R 0 CLUSTER SETSLOT 609 MIGRATING $R1_id TIMEOUT 99999999999999999999} e
+        assert_equal $e "ERR timeout is not an integer or out of range"
+
+        catch {R 0 CLUSTER SETSLOT 609 MIGRATING $R1_id TIMEOUT abc} e
+        assert_equal $e "ERR timeout is not an integer or out of range"
+
+        catch {R 0 CLUSTER SETSLOT 609 TIMEOUT 100 MIGRATING $R1_id} e
+        assert_equal $e "ERR Invalid CLUSTER SETSLOT action or number of arguments. Try CLUSTER HELP"
+    }
+}
```
Comment (zuiderkwast) [inline tests/unit/cluster/slot-migration.tcl:401]: It's good to get coverage on all these cases, but TBH I think we should use `addReplyErrorObject(c, shared.syntaxerr);` in most of these cases.

**Review by zuiderkwast — COMMENTED** [review]
You're right. I see now in the docs for BLPOP:

> A timeout of zero can be used to block indefinitely.

Let's be explicit in the title of the PR that this is what it means.

Just some minor comment about syntax error reply. Otherwise LGTM.

File: src/cluster_legacy.c (hunk 1 above)
Comment (PingXie) [reply src/cluster_legacy.c:5969]: The syntax error string is too generic. Also `getTimeoutFromObjectOrReply()` returns specific errors too. I think it makes sense to go along with that.

File: tests/unit/cluster/slot-migration.tcl (hunk 2 above)
Comment (PingXie) [reply tests/unit/cluster/slot-migration.tcl:401]: My preference is to align with `getTimeoutFromObjectOrReply()`.

File: src/cluster_legacy.c (hunk 1 above)
Comment (zuiderkwast) [reply src/cluster_legacy.c:5969]: OK, I accept you preference. :-)

My thinking was that there's enough syntax information available in our command JSON files to let clients validate the syntax on the client side. Our errors don't need to be verbose docs.

**Review by zuiderkwast — APPROVED** [review]

File: src/cluster_legacy.c (hunk 1 above)
Comment (PingXie) [reply src/cluster_legacy.c:5969]: I would agree with you if it was not for getTimeoutFromObjectOrReply :)

**Pull request merged**